mqtt/mqtt-bridge/tests/integration.rs

```diff
@@ -0,0 +1,418 @@
+        .with_client_id(ClientId::IdWithExistingSession("upstream_client".into()))
+        .build();
+
+    upstream_client
+        .subscribe("upstream/temp/#", QoS::AtLeastOnce)
+        .await;
+    // wait to receive subscription ack
+    upstream_client.subscriptions().recv().await;
+
+    // Send upstream
+    local_client
+        .publish_qos1("to/temp/1", "from local", false)
+        .await;
+
+    // Send downstream
+    upstream_client
+        .publish_qos1("to/filter/1", "from upstream", false)
+        .await;
+
+    assert_matches!(
+        upstream_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from local")
+    );
+
+    assert_matches!(
+        local_client.publications().recv().await,
+        Some(ReceivedPublication{payload, .. }) if payload == Bytes::from("from upstream")
+    );
+
+    upstream_client.shutdown().await;
+    local_client.shutdown().await;
+}
+
+#[tokio::test]
+async fn bridge_settings_update() {
+    let (local_server_handle, upstream_server_handle, _) = setup_brokers(AllowAll, AllowAll);
+    let mut controller_handle = setup_bridge_controller(
+        local_server_handle.address(),
+        upstream_server_handle.tls_address().unwrap(),
+        vec![],
+    )
+    .await;
+
+    let mut local_client = TestClientBuilder::new(local_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("local_client".into()))
+        .build();
+
+    local_client
+        .subscribe("downstream/filter/#", QoS::AtLeastOnce)
+        .await;
+    // wait to receive subscription ack
+    local_client.subscriptions().recv().await;
+
+    let mut upstream_client = TestClientBuilder::new(upstream_server_handle.address())
+        .with_client_id(ClientId::IdWithExistingSession("upstream_client".into()))
+        .build();
+
+    upstream_client
+        .subscribe("upstream/temp/#", QoS::AtLeastOnce)
+        .await;
+    // wait to receive subscription ack
+    upstream_client.subscriptions().recv().await;
+
+    // Send upstream
+    local_client
+        .publish_qos1("to/temp/1", "from local before update", false)
+        .await;
+
+    // Send downstream
+    upstream_client
+        .publish_qos1("to/filter/1", "from upstream before update", false)
+        .await;
+
+    let subs = vec![TopicRule::new(
+        "filter/#".into(),
+        Some("to".into()),
+        Some("downstream".into()),
+    )];
+    let forwards = vec![TopicRule::new(
+        "temp/#".into(),
+        Some("to".into()),
+        Some("upstream".into()),
+    )];
+
+    controller_handle
+        .send_update(BridgeControllerUpdate::from_bridge_topic_rules(
+            "$upstream",
+            subs.as_ref(),
+            forwards.as_ref(),
+        ))
+        .unwrap();
+
+    // delay to propagate the update
+    tokio::time::delay_for(Duration::from_secs(2)).await;
```

Comment:
If we can use signals here instead of a wait that will be better, but I don't think it is possible.

Reply:
@vadim-kovalyov Do you remember implementing awaiting for a signal?

Reply:
I implemented awaiting for an authorizer updates. This is different.